### PR TITLE
Don't panic in ceil_char_boundary

### PR DESCRIPTION
**Pull request by clarfonthey**
Implementing the alternative mentioned in this comment: https://github.com/rust-lang/rust/issues/93743#issuecomment-1579935853

Since `floor_char_boundary` will always work (rounding down to the length of the string is possible), it feels best for `ceil_char_boundary` to not panic either. However, the semantics of "rounding up" past the length of the string aren't very great, which is why the method originally panicked in these cases.

Taking into account how people are using this method, it feels best to simply return the end of the string in these cases, so that the result is still a valid char boundary.

### CODE REVIEW

File: library/core/src/str/mod.rs [hunk 1]
```diff
@@ -271,14 +271,13 @@ impl str {
 
     /// Finds the closest `x` not below `index` where `is_char_boundary(x)` is `true`.
     ///
+    /// If `x` is greater than the length of the string, this returns the length of the string.
```
Comment (cuviper) [inline library/core/src/str/mod.rs:274]: I suppose you mean:
```suggestion
    /// If `index` is greater than the length of the string, this returns the length of the string.
```

**Comment by cuviper** [comment]
I think this needs to be an API question, even though it's not stable yet. It seems surprising to me that it would return something less than `index` when it is stated up front, "not below `index`", so panicking is the only choice. But I also kind of think that floor should not paper over out-of-bounds `index` either.

@bors label -T-libs +T-libs-api
r? libs-api

**Comment by clarfonthey** [comment]
For a more direct link to what I was searching: https://github.com/search?utf8=%E2%9C%93&q=ceil_char_boundary+language%3ARust&type=code&l=Rust

Essentially, I saw a few uses of `ceil_char_boundary(index + 1)` or something to that effect, and while I didn't scrutinise the exact usages to make sure they were using it properly, that alongside the combination of people asking for a non-panicking version made me feel that it would make the most sense for this to not panic.

My (flimsy) logic is to think of it like climbing a ladder, and if you climb past the last rung on the ladder you just stay at the last rung.

**Comment by m-ou-se** [comment]
We discussed this in the libs-api meeting last week. While we felt it was a bit odd for this method to return a value lower than its input, we agree that one of its main use cases is basically to sanitize input to e.g. `s.split_at(x)`, `s[..x]`, etc. The guarantee that any return value from `floor_char_boundary` and `ceil_char_boundary` is always a valid index to give to `split_at` is a very useful one.

Since this is still unstable, this PR doesn't require an FCP. (The behavior of this method will be part of the FCP that will happen when stabilizing it.)

@bors r+

**Pull request merged**